### ISSUE TITLE
Generalised gradient descent algorithm 

### ISSUE THREAD

**Issue by jfrancis71** [jfrancis71/Machine-Vision#14]
Presently gradient descent is implemented specifically for optimising neural net parameters and it is also implemented for the Dream code.

Is it possible to have single code base for both? Maybe this way we could automatically have things like Nesterov and classical momentum implemented for both in a straightforward fashion without lots of duplicated code,
